diffkemp/simpll/InstPatternComparator.cpp

```diff
@@ -0,0 +1,655 @@
+//===---- InstPatternComparator.cpp - Code pattern instruction matcher ----===//
+//
+//       SimpLL - Program simplifier for analysis of semantic difference      //
+//
+// This file is published under Apache 2.0 license. See LICENSE for details.
+// Author: Petr Silling, psilling@redhat.com
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file contains the implementation of the LLVM code pattern matcher. The
+/// pattern matcher is a comparator extension of the LLVM FunctionComparator
+/// tailored to comparison of general instruction-based patterns.
+///
+//===----------------------------------------------------------------------===//
+
+#include "InstPatternComparator.h"
+#include "Config.h"
+#include "Utils.h"
+
+/// Compare the module function and the difference pattern from the starting
+/// module instruction. This includes checks for correct input mappings.
+int InstPatternComparator::compare() {
+    // Clear all previous results.
+    beginCompare();
+    InstMatchMap.clear();
+    InputMatchMap.clear();
+    ReverseInputMatchMap.clear();
+
+    // Run the main matchin algorithm.
+    if (int Res = matchPattern())
+        return Res;
+
+    // Reset the comparison state without clearing pattern matches.
+    beginCompare();
+
+    // Ensure that the created input mapping is correct. All input instructions
+    // and arguments have to be mapped correctly.
+    if (int Res = checkInputMapping())
+        return Res;
+
+    return 0;
+}
+
+/// Set the starting module instruction.
+void InstPatternComparator::setStartInstruction(
+        const Instruction *StartModInst) {
+    StartInst = StartModInst;
```

Comment:
Not really. I was sceptical about this one as well. I simply feel that `StartInst` should not be a public attribute. What is your opinion?

Reply:
Hm, maybe makes sense not to have it public but since we're not using explicit setters anywhere, I'd rather stick to that behaviour. That is, remove the setter.